### PR TITLE
Carry back remote call error

### PR DESCRIPTION
**Pull request by yim-lee**
Resolves https://github.com/apple/swift-distributed-actors/issues/887

### CODE REVIEW

File: Sources/DistributedActors/ClusterSystem.swift [hunk 1]
```diff
@@ -1019,20 +1029,98 @@ public struct ClusterInvocationResultHandler: DistributedTargetInvocationResultH
...
+            let errorManifest = try context.serialization.outboundManifest(type(of: thrownError))
+            try container.encode(thrownError, forKey: .thrownError)
+            try container.encode(errorManifest, forKey: .thrownErrorManifest)
+        } else {
+            try container.encode(self.value, forKey: .value)
+        }
+    }
+}
+
+public struct GenericRemoteCallError: Error, Codable {
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:1122]: Cool excellent :)

File: Sources/DistributedActors/Refs.swift [hunk 2]
```diff
@@ -237,6 +237,19 @@ extension _ActorRef {
...
+            // The guard check above fails for the latter use-case because of message type mismatch.
+            // The logic below converts the error reply to the proper type then sends it to the actor.
+            if let ReplyType = Message.self as? any AnyRemoteCallReply.Type,
+               let remoteCallReply = message as? any AnyRemoteCallReply,
+               let remoteCallError = remoteCallReply.thrownError,
+               let _message = ReplyType.init(error: remoteCallError) as? Message
+            {
+                self.tell(_message, file: file, line: line)
+                return
+            }
```
Comment (ktoso) [inline Sources/DistributedActors/Refs.swift:251]: ok this is fine for the current state of the runtime 👍

File: Sources/DistributedActors/ClusterSystem.swift [hunk 3]
```diff
@@ -937,7 +937,14 @@ extension ClusterSystem {
             return invocation
         }
 
-        return try await ask.value
+        let reply = try await ask.value
+        if let error = reply.thrownError {
+            throw error
+        }
+        guard let value = reply.value else {
+            fatalError("Invalid RemoteCallReply, expected either value or thrownError to be set.")
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:945]: Let's just throw here -- make some new RemoteCallError case for it; I would not want to allow our system to crash only if someone managed to send us a bad message like that 👍

File: Sources/DistributedActors/ClusterSystem.swift [hunk 4]
```diff
@@ -937,7 +937,14 @@ extension ClusterSystem {
...
+        }
+        return value
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:947]: Pretty nice how the remoteCall impls are pretty neat and clean now :)

File: Sources/DistributedActors/ClusterSystem.swift [hunk 5]
```diff
@@ -1019,20 +1029,98 @@ public struct ClusterInvocationResultHandler: DistributedTargetInvocationResultH
...
         self.system.log.warning("Result handler, onThrow: \(error)")
-        // FIXME(distributed): carry the error back
-        fatalError("FIXME: implement sending back errors")
+        let ref = _ActorRef<RemoteCallReply<_Done>>(.remote(.init(shell: clusterShell, address: replyToAddress, system: system)))
+        if let codableError = error as? (Error & Codable) {
+            ref.tell(.init(error: codableError))
+        } else {
+            ref.tell(.init(error: GenericRemoteCallError(message: "Remote call error of [\(type(of: error as Any))] type occurred")))
+        }
+    }
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:1049]: Awesome 👍 

I think we may want to indeed do some configuration here... some "default remote call error handling" and set it to `.genericRemoteCallError` or `.opaqueError` if we'd want to not expose the exact error that was thrown but just "some error happened" I guess?

That's a separate ticket though and not for beta1 :nod:

File: Sources/DistributedActors/ClusterSystem.swift [hunk 6]
```diff
@@ -1019,20 +1029,98 @@ public struct ClusterInvocationResultHandler: DistributedTargetInvocationResultH
...
+}
+
+protocol AnyRemoteCallReply: Codable {
+    associatedtype Value: Codable
+
+    var value: Value? { get }
+    var thrownError: (any Error & Codable)? { get }
+
+    init(value: Value)
+    init<Err: Error & Codable>(error: Err)
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:1059]: do we need the inits in the protocol? I'd expect not to need them, and only need the getters 🤔

File: Tests/DistributedActorsTests/ClusterSystemTests.swift [hunk 7]
```diff
@@ -202,14 +180,153 @@ final class ClusterSystemTests: ActorSystemXCTestCase {
...
+        }
+
+        let noncodableError = try await shouldThrow {
+            _ = try await remoteGreeter.helloThrow(codable: false)
+        }
+        guard let genericRemoteCallError = noncodableError as? GenericRemoteCallError else {
+            throw testKit.fail("Expected GenericRemoteCallError, got \(noncodableError)")
+        }
+        genericRemoteCallError.message.shouldStartWith(prefix: "Remote call error of [GreeterNonCodableError] type occurred")
+    }
```
Comment (ktoso) [inline Tests/DistributedActorsTests/ClusterSystemTests.swift:217]: Can we put the throwing cases into their own tests?

```
func test_remoteCall_shouldCarryBackThrownError()
```

?

**Review by ktoso — APPROVED** [review]
LGTM, awesome! Worked out quite well, just minor comments and merge at will 👍

File: Sources/DistributedActors/ClusterSystem.swift (hunk 6 above)
Comment (yim-lee) [reply Sources/DistributedActors/ClusterSystem.swift:1059]: It's used in [`_tellOrDeadLetter`](https://github.com/apple/swift-distributed-actors/pull/931/files#diff-19182eaca69e6caa9693cf4f221489ea4d4d1cde2644dad95726066984189aaaR247) to construct new `RemoteCallReply`. Open to suggestions.

Comment (ktoso) [reply Sources/DistributedActors/ClusterSystem.swift:1059]: Oh I see! Super sneaky, I love it -- good to keep 👍

File: Sources/DistributedActors/ClusterSystem.swift (hunk 5 above)
Comment (yim-lee) [reply Sources/DistributedActors/ClusterSystem.swift:1049]: Created https://github.com/apple/swift-distributed-actors/issues/932

**Comment by ktoso** [comment]
LGTM again, thank you!

**Pull request merged**